File: kit/android/ble-payment-kit/src/main/kotlin/ru/paymentguide/blepaymentkit/AdvertisementPacketParser.kt
package ru.paymentguide.blepaymentkit

data class AdvertisementPacket(
    val packetVersion: Int,
    val rssiDelta: Int,
    val terminalCapabilities: Int,
    val operationCounter: Int,
    val qrcId: String,
    val qrcPayload: ByteArray,
) {
    override fun equals(other: Any?): Boolean = other is AdvertisementPacket &&
        packetVersion == other.packetVersion &&
        rssiDelta == other.rssiDelta &&
        terminalCapabilities == other.terminalCapabilities &&
        operationCounter == other.operationCounter &&
        qrcId == other.qrcId &&
        qrcPayload.contentEquals(other.qrcPayload)

    override fun hashCode(): Int {
        var result = packetVersion
        result = 31 * result + rssiDelta
        result = 31 * result + terminalCapabilities
        result = 31 * result + operationCounter
        result = 31 * result + qrcId.hashCode()
        result = 31 * result + qrcPayload.contentHashCode()
        return result
    }
}

class AdvertisementPacketParser {
    fun parse(data: ByteArray, config: BlePaymentConfig): Result<AdvertisementPacket> = runCatching {
        if (data.size < 3) throw PacketRejected(BlePacketRejectReason.packetTooShort)
        val versionAndDelta = data[0].toInt() and 0xFF
        val packetVersion = (versionAndDelta shr 5) and 0b111
        if (packetVersion != config.supportedPacketVersion) throw PacketRejected(BlePacketRejectReason.unsupportedVersion)
        val deltaBits = versionAndDelta and 0b1_1111
        val rssiDelta = if ((deltaBits and 0b1_0000) != 0) deltaBits - 0b10_0000 else deltaBits
        val capabilities = data[1].toInt() and 0xFF
        if ((capabilities and config.requiredCapabilityMask) != config.requiredCapabilityMask) {
            throw PacketRejected(BlePacketRejectReason.missingRequiredField)
        }
        val payload = data.copyOfRange(3, data.size)
        AdvertisementPacket(packetVersion, rssiDelta, capabilities, data[2].toInt() and 0xFF, ByteUtils.base36UnsignedBigEndian(payload), payload)
    }
}

File: ios/BlePaymentKit/Sources/BlePaymentKit/AdvertisementPacketParser.swift
import Foundation

public struct AdvertisementPacket: Equatable, Sendable {
    public let packetVersion: UInt8
    public let rssiDelta: Int
    public let capabilities: UInt8
    public let operationCounter: UInt8
    public let qrcId: String
    public let qrcPayload: Data
}

public struct AdvertisementPacketParser: Sendable {
    public init() {}

    public func parse(_ data: Data, config: BlePaymentConfig) -> Result<AdvertisementPacket, BlePacketRejectReason> {
        guard data.count >= 3 else { return .failure(.packetTooShort) }

        let versionAndDelta = data[0]
        let packetVersion = (versionAndDelta >> 5) & 0b111
        guard packetVersion == config.supportedPacketVersion else {
            return .failure(.unsupportedVersion)
        }

        let deltaBits = Int(versionAndDelta & 0b1_1111)
        let rssiDelta = (deltaBits & 0b1_0000) != 0 ? deltaBits - 0b10_0000 : deltaBits

        let capabilities = data[1]
        guard (capabilities & config.requiredCapabilityMask) == config.requiredCapabilityMask else {
            return .failure(.missingRequiredField)
        }

        let operationCounter = data[2]
        let qrcPayload = data.count > 3 ? Data(data.dropFirst(3)) : Data()
        let qrcId = ByteUtils.base36UnsignedBigEndian(qrcPayload)

        return .success(AdvertisementPacket(packetVersion: packetVersion, rssiDelta: rssiDelta, capabilities: capabilities, operationCounter: operationCounter, qrcId: qrcId, qrcPayload: qrcPayload))
    }
}
